### ISSUE TITLE
improve go generate support

### ISSUE THREAD

**Issue by blaggacao** [vektra/mockery#272]
Having to do such thing is annoying:

```shell
➜  git git:(da-git-config-mock) ✗ go generate -n
docker run -v /home/blaggacao/go/src/github.com/git-town/git-town/src/git/../..:/src -w /src --user=$(`id-u`) vektra/mockery -dir=src/git -name ConfigurationInterface
➜  git git:(da-git-config-mock) ✗ go generate   
01 Jun 20 07:07 UTC INF Starting mockery version=2.0.0-alpha.2
01 Jun 20 07:07 UTC INF Walking version=2.0.0-alpha.2
01 Jun 20 07:07 UTC INF Generating mock interface=ConfigurationInterface path=mocks/ConfigurationInterface.go version=2.0.0-alpha.2
➜  git git:(da-git-config-mock) ✗ ls ../../mocks 
ConfigurationInterface.go
➜  git git:(da-git-config-mock) ✗ ls -la ../../mocks
total 32
drwxrwxr-x  2 blaggacao blaggacao  4096 Jun  1 02:07 .
drwxrwxr-x 16 blaggacao blaggacao  4096 Jun  1 01:39 ..
-rw-r--r--  1 root      root      16121 Jun  1 02:07 ConfigurationInterface.go
➜  git git:(da-git-config-mock) ✗ sudo chown $(id -u):$(id -g) -R ../../mocks
[sudo] password for blaggacao
```

while `$(id -u):$(id -g)` unfortunately isn't interpreted,
```shell
➜  git git:(da-git-config-mock) ✗ go generate -n                         
docker run -v /home/blaggacao/go/src/github.com/git-town/git-town/src/git/../..:/src -w /src --user=$(id -u):$(id -g) vektra/mockery -dir=src/git -name ConfigurationInterface
```

**Comment by blaggacao** [comment]
Oh not even with a shell indirection:
```shell
sh -c docker run -v /home/blaggacao/go/src/github.com/git-town/git-town/src/git/../..:/src -w /src --user=$(id -u):$(id -g) vektra/mockery -dir=src/git -name ConfigurationInterface
➜  git git:(da-git-config-mock) ✗ go generate                            
01 Jun 20 07:11 UTC INF Starting mockery version=2.0.0-alpha.2
01 Jun 20 07:11 UTC INF Walking version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC ERR Error parsing file error="err: exit status 1: stderr: failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied\n" version=2.0.0-alpha.2
01 Jun 20 07:11 UTC FTL Unable to find 'ConfigurationInterface' in any go files under this path version=2.0.0-alpha.2
config.go:21: running "sh": exit status 1
```

**Comment by blaggacao** [comment]
Ok, this works, but isn't it strange with sudo?
```shell
//go:generate docker run -v $PWD/../..:/src -w /src vektra/mockery -dir=src/git -name ConfigurationInterface
//go:generate echo "\nFiles are generated within the user namespace of the\ncontainer which is mapped to root. Therefore we sudo chown unconventionally...\n"
//go:generate sh -c "echo sudo chown $(id -u):$(id -g) -R ../../mocks\n"
//go:generate sh -c "sudo chown $(id -u):$(id -g) -R ../../mocks"
```

but it's not so elegant.

**Comment by LandonTClipp** [comment]
So I'm not really sure what the solution would be for this, you are generating this with `root` inside the docker image. If you want the files to be owned by something else, you need to be running your container under a non-root user. That's not something that can be fixed in this project because we can't know which UID you want to run as.

**Comment by blaggacao** [comment]
What _could_ be  done is write to std-out and pipe it into the file, but that has other issues with it regarding folder layout. At least it would allow to use a tagged version of mockery for any given repo, which I wouldn't know how to realiably do with the binary.

**Comment by LandonTClipp** [comment]
I'm curious if there could be some kind of workaround we can bake into mockery itself, like perhaps via configuration you can tell it to chown to a certain UID/GID after writing it out.

**Comment by blaggacao** [comment]
Me reflex was to use `--user=$(id -u):$(id -g)` to run mockery as my host user, but then mockery docker image bailed not having permission. https://github.com/vektra/mockery/issues/272#issuecomment-636656601

If that could be solved, then it should work wrapping the go generate command into a `sh -c`...
That would be indeed the ideal solution.

**Comment by blaggacao** [comment]
@LandonTClipp Could you put a makefile command to quickly build the docker image if mockery binary is not yet present? I messed it up somehow, and now the binary is in a completely different path...

**Comment by LandonTClipp** [comment]
@blaggacao do you still want that Makefile change? I'm pretty sure you fixed that in your PR #280?

**Comment by blaggacao** [comment]
Yap, thanks I fixed it on the go.
